themes/default/content/blog/easy-langserve-apps-with-pulumi/index.md

```diff
@@ -0,0 +1,440 @@
+    raise Exception("Missing `PINECONE_API_KEY` environment variable.")
+
+PINECONE_INDEX_NAME = os.environ.get("PINECONE_INDEX", "test")
+
+### Ingest code - you may need to run this the first time
+# Load
+from langchain_community.document_loaders import WikipediaLoader
+
+loader = WikipediaLoader(query="Gandalf the Grey", load_max_docs=50)
+data = loader.load()
+
+# # Split
+from langchain.text_splitter import RecursiveCharacterTextSplitter
+
+text_splitter = RecursiveCharacterTextSplitter(chunk_size=500, chunk_overlap=0)
+all_splits = text_splitter.split_documents(data)
+
+# # Add to vectorDB
+vectorstore = Pinecone.from_documents(
+    documents=all_splits, embedding=OpenAIEmbeddings(), index_name=PINECONE_INDEX_NAME
+)
+retriever = vectorstore.as_retriever()
+### End Ingest code
+
+vectorstore = Pinecone.from_existing_index(PINECONE_INDEX_NAME, OpenAIEmbeddings())
+retriever = vectorstore.as_retriever()
+
+# RAG prompt
+template = """Answer the question based only on the following context:
+{context}
+Question: {question}
+"""
+prompt = ChatPromptTemplate.from_template(template)
+
+# RAG
+model = ChatOpenAI()
+chain = (
+    RunnableParallel({"context": retriever, "question": RunnablePassthrough()})
+    | prompt
+    | model
+    | StrOutputParser()
+)
+
+
+# Add typing for input
+class Question(BaseModel):
+    __root__: str
+
+
+chain = chain.with_types(input_type=Question)
+```
+
+Head over to the `server.py` file and add the following code:
+
+```python
+from app.chain import chain as rag_chain
+
+# ... rest of the code
+add_routes(
+    app,
+    rag_chain,
+    path="/rag-pinecone",
+)
+```
+
+To deploy the chain with Pulumi, you can use the following command:
+
+```shell
+pulumi login # if you haven't logged in yet
+pulumi config set open-api-key --secret # your openai api key
+pulumi config set pinecone-api-key --secret # your pinecone api key
+pulumi up
+```
+
+This will deploy create the AWS infrastructure, and the Pinecone index and then containerize and deploy the LangServe
+app to AWS Fargate.
+
+After a couple of minutes, you will see in the Pinecone console that the index has been created and the first documents
+have been ingested from Wikipedia about Gandalf the Grey.
+
+Here a screenshot of the Pinecone console:
+
+![Pinecone Console](./pinecone-console.png)
+
+And if you want to test the LangServe app, you can use the playground UI, where you should be able to ask questions
+about Gandalf the Grey and get answers based on the documents ingested into the Pinecone index.
+
+![LangServe Pinecone](./langserve-pinecone.png)
+
+If you want to destroy the stack, you can use the following command:
+
+```shell
+pulumi destroy
+```
+
+The Pulumi CLI will ask you to confirm the changes, and if you are happy with the changes, you can confirm them by
+typing `yes` and all the resources will be deleted, including the Pinecone index.
+
+All the resources will be deleted, including the Pinecone index.
```

Comment:
Seems redundant with the previous paragraph?

Reply:
so true! 😅